vignettes/rcppannoy.Rmd

```diff
@@ -10,6 +10,13 @@ vignette: >
   %\usepackage[utf8]{inputenc}
 ---
 
+```{r, echo=FALSE, results="hide"}
+knitr::opts_chunk$set(error=FALSE, warning=FALSE, message=FALSE)
+```
+
+```{Rcpp, ref.label=knitr::all_rcpp_labels(), include=FALSE}
+```
```

Comment:
What does this too?   Possibly not enough coffee here...

Reply:
This is the cool part - https://bookdown.org/yihui/rmarkdown/language-engines.html#rcpp, which takes all the Rcpp chunks in the document and compiles them in a single unit.

Reply:
But I don't think you need the label.  Based on the 100+ files we have at the Rcpp Gallery this has always been the case.  (Or maybe I confuse Gallery posts written in .cpp with those in .Rmd.)

Reply:
It seems that there's two effects here. The obvious one is that, if the `ref.label=` is removed, the C++ compilation is skipped entirely (because all of the Rcpp chunks are `eval=FALSE`). The more subtle effect of the `ref.label=` is that all of the individual C++ blocks are concatenated into a single chunk of code for compilation at once. If we `eval=TRUE` all of the Rcpp chunks but remove the `ref.label=`, _knitr_ tries to compile each individual Rcpp chunk in the vignette separately, resulting in a bunch of errors as you might expect. I can't compare to what happens in the Rcpp gallery, I've never posted anything there, but a cursory inspection of some of the gallery posts suggests that each of the C++ blocks define complete Rcpp functions and can be compiled separately with no problems.

Reply:
Good reasoning.  I _think_ the Rmd did in fact do it piece-meal which made cross-reference across chunks harder / impossible.  The `.cpp` did it all as one ODR / compilation unit.   So in that sense this is a good trick, and thanks for reaching me about it :)

Now, do we _need_ compilation in the vignette?  I'd actually say no but am open to be convinced otherwise.

Reply:
I would say that compilation of the vignette in some manner is helpful, if only to ensure that the C++ code that we recommend to users is actually correct (for a given version of the Annoy libraries, and across different systems). Or at least compiles. The first version of the vignette contained at least two errors (#31, and looping across columns instead of rows) and a bunch of not-so-clear omissions (e.g., failing to `#include "Rcpp.h"`; failing to mention where `std::copy` comes from; lapsing into C++11). These were pretty easy to identify and fix once I tried to compile the entire thing.

Of course, there's no need to actually do the compilation in the vignette itself. One can imagine having the script mentioned in #31 as a unit test. But having the compilation in the vignette was the cleanest way to do it; actually inspired by your mention of ` ```{Rcpp}`, which I didn't know about before.